compiler/compiler/src/lib.rs

```diff
@@ -73,24 +74,24 @@ impl<'a> Compiler<'a> {
         Ok(format!("{:x}", hash))
     }
 
-    ///
-    /// Runs the compiler stages.
-    ///
-    fn compiler_stages(self) -> Result<leo_ast::Ast> {
-        //load the input file if it exists.
-        let _input_ast = if self.input_file_path.exists() {
-            let input_string = fs::read_to_string(&self.input_file_path)
-                .map_err(|e| CompilerError::file_read_error(self.main_file_path.clone(), e))?;
-
-            Some(leo_parser::parse_input(
-                self.handler,
-                self.input_file_path.to_str().unwrap_or_default(),
-                input_string,
-            )?)
-        } else {
-            None
-        };
+    // Parses and stores a program file content from a string, constructs a syntax tree, and generates a program.
+    pub fn parse_program_from_string(&mut self, program_string: &str) -> Result<()> {
+        // Use the parser to construct the abstract syntax tree (ast).
+        let ast: leo_ast::Ast = leo_parser::parse_ast(
+            self.handler,
+            self.main_file_path.to_str().unwrap_or_default(),
+            program_string,
+        )?;
+        // Write the AST snapshot post parsing.
+        ast.to_json_file_without_keys(self.output_directory.clone(), "initial_ast.json", &["span"])?;
+
+        self.ast = ast;
+
+        Ok(())
+    }
 
+    /// Parses and stores the main program file, constructs a syntax tree, and generates a program.
+    pub fn parse_program(&mut self) -> Result<()> {
```

Comment:
This method should call `parse_program_from_string` now that the logic is the same.